src/middlewared/middlewared/plugins/kmip.py

```diff
@@ -0,0 +1,384 @@
+                [ca['certificate']]
+            ):
+                verrors.add(
+                    'kmip_update.certificate_authority',
+                    'Certificate chain could not be verified with specified certificate authority.'
+                )
+        elif not ca:
+            verrors.add('kmip_update.certificate_authority', 'Please specify a valid id.')
+
+        if new.pop('validate', True) and new['enabled'] and not verrors:
+            result = await self.middleware.run_in_thread(self.test_connection, new)
+            if result['error']:
+                verrors.add('kmip_update.server', f'Unable to connect to KMIP server: {result["exception"]}.')
+
+        force_clear = new.pop('force_clear', False)
+        sync_error = 'KMIP sync is pending, please make sure database and KMIP server ' \
+                     'are in sync before proceeding with this operation.'
+        if old['enabled'] != new['enabled'] and await self.kmip_sync_pending():
+            if force_clear:
+                await self.clear_sync_pending_keys()
+            else:
+                verrors.add('kmip_update.enabled', sync_error)
+
+        verrors.check()
+
+        await self.middleware.call(
+            'datastore.update', self._config.datastore, old['id'], new,
+        )
+
+        await self.middleware.call('service.start', 'kmip')
+        if new['enabled'] and old['enabled'] != new['enabled']:
+            await self.middleware.call('kmip.initialize_keys')
+        if any(old[k] != new[k] for k in ('enabled', 'manage_zfs_keys', 'manage_sed_keys')):
+            await self.middleware.call('kmip.sync_keys')
+
+        return await self.config()
+
+    @private
+    def initialize_zfs_keys(self):
+        for ds in self.middleware.call_sync(
+            'datastore.query', self.middleware.call_sync('pool.dataset.dataset_datastore')
+        ):
+            if ds['encryption_key']:
+                self.zfs_keys[ds['name']] = self.middleware.call_sync('pwenc.decrypt', ds['encryption_key'])
+            elif ds['kmip_uid']:
+                try:
+                    key = self.retrieve_secret_data(ds['kmip_uid'])
+                except Exception:
+                    self.middleware.logger.debug(f'Failed to retrieve key for {ds["name"]}')
+                else:
+                    self.zfs_keys[ds['name']] = key
+
+    @private
+    @job(lock='initialize_kmip_keys')
+    def initialize_keys(self, job):
+        kmip_config = self.middleware.call_sync('kmip.config')
+        if kmip_config['manage_zfs_keys']:
+            self.initialize_zfs_keys()
+
+    @private
+    async def retrieve_zfs_keys(self):
+        return self.zfs_keys
+
+
+async def __event_system(middleware, event_type, args):
+    if args['id'] != 'ready':
+        return
+
+    if (await middleware.call('kmip.config'))['enabled']:
+        await middleware.call('kmip.initialize_keys')
+
```

Comment:
Please look how I've split the datastore plugin into few files and try to do the same with this file. Long files are pain to maintain. I can see at least three parts:
* part that interacts with kmip library (connection context manager and operations)
* part that syncs keys
* part that implements ConfigService